Remote sync fails if temp directory is deleted, until instance restart

### Describe the bug

If the directory `$TMPDIR/metabase-git` is deleted, all remote sync operations will fail with errors like `origin: not found`. The only way to recover is to restart the instance, which will clone the local repository again.

### To Reproduce

1. Set up remote sync
2. Run `rm -rf /tmp/metabase-git` (or wherever your temp directory is)
3. Try any remote sync operation (push, pull or switch branch)
4. See error


### Expected behavior

Although the official container image doesn't automatically delete temp files, we have to guard against them not existing for whatever reason. All remote sync operations should still work even if the repo needs to be cloned again.

### Logs

```
2026-03-24 07:17:04,255 ERROR remote-sync.api :: Failed to get branches from source: Git LsRemoteCommand failed: origin: not found. 
clojure.lang.ExceptionInfo: Git LsRemoteCommand failed: origin: not found.
	at metabase_enterprise.remote_sync.source.git$clean_git_exception.invokeStatic(git.clj:33)
	at metabase_enterprise.remote_sync.source.git$clean_git_exception.invoke(git.clj:29)
	at metabase_enterprise.remote_sync.source.git$call_remote_command.invokeStatic(git.clj:59)
	at metabase_enterprise.remote_sync.source.git$call_remote_command.invoke(git.clj:46)
	at metabase_enterprise.remote_sync.source.git$branches.invokeStatic(git.clj:348)
	at metabase_enterprise.remote_sync.source.git$branches.invoke(git.clj:340)
	at metabase_enterprise.remote_sync.source.git.GitSource.branches(git.clj:473)
	at metabase_enterprise.remote_sync.api$fn__164117$fn__164123$fn__164124.invoke(api.clj:195)
	at metabase_enterprise.remote_sync.api$fn__164117$fn__164123.invoke(api.clj:194)
	at metabase.api.macros$_core_fn$core_fn__102927.invoke(macros.clj:426)
	at metabase.api.macros$endpoint_handler_STAR_$handler__102969.invoke(macros.clj:622)
	at metabase.api.macros$build_ns_handler$ns_handler_STAR___102995.invoke(macros.clj:696)
	at metabase.api.open_api.HandlerWithOpenAPISpec.invoke(open_api.clj:42)
	at metabase.api.routes.common$enforce_authentication$fn__103414.invoke(common.clj:92)
	at metabase.api.open_api.HandlerWithOpenAPISpec.invoke(open_api.clj:42)
	at metabase_enterprise.api.routes.common$_PLUS_require_premium_feature$fn__156591.invoke(common.clj:23)
	at metabase.api.open_api.HandlerWithOpenAPISpec.invoke(open_api.clj:42)
	at metabase.api.util.handlers$_route_map_handler$fn__103327.invoke(handlers.clj:23)
	at metabase.api.open_api.HandlerWithOpenAPISpec.invoke(open_api.clj:42)
	at metabase.api.util.handlers$_route_map_handler$fn__103327.invoke(handlers.clj:23)
	at metabase.api.open_api.HandlerWithOpenAPISpec.invoke(open_api.clj:42)
	at clojure.lang.Var.invoke(Var.java:395)
	at compojure.core$routes$fn__103264$f__103265.invoke(core.clj:198)
	at compojure.core$routes$fn__103264.invoke(core.clj:200)
	at metabase.api.open_api.HandlerWithOpenAPISpec.invoke(open_api.clj:42)
	at clojure.lang.Var.invoke(Var.java:395)
	at metabase.server.routes$api_handler$api_handler_STAR___109654.invoke(routes.clj:88)
	at compojure.core$routes$fn__103264$f__103265.invoke(core.clj:198)
	at compojure.core$routes$fn__103264.invoke(core.clj:200)
	at compojure.core$make_context$handler__103292.invoke(core.clj:290)
	at compojure.core$make_context$fn__103296.invoke(core.clj:300)
	at compojure.core$routes$fn__103264$f__103265.invoke(core.clj:198)
	at compojure.core$routes$fn__103264$f__103265$respond_SINGLEQUOTE___103266.invoke(core.clj:197)
	at compojure.core$wrap_route_matches$fn__103245.invoke(core.clj:153)
	at compojure.core$routes$fn__103264$f__103265.invoke(core.clj:198)
	at compojure.core$routes$fn__103264$f__103265$respond_SINGLEQUOTE___103266.invoke(core.clj:197)
	at compojure.core$wrap_route_matches$fn__103245.invoke(core.clj:153)
	at compojure.core$routes$fn__103264$f__103265.invoke(core.clj:198)
	at compojure.core$routes$fn__103264$f__103265$respond_SINGLEQUOTE___103266.invoke(core.clj:197)
	at compojure.core$wrap_route_matches$fn__103245.invoke(core.clj:152)
	at compojure.core$routes$fn__103264$f__103265.invoke(core.clj:198)
	at compojure.core$routes$fn__103264$f__103265$respond_SINGLEQUOTE___103266.invoke(core.clj:197)
	at compojure.core$wrap_route_matches$fn__103245.invoke(core.clj:152)
	at compojure.core$routes$fn__103264$f__103265.invoke(core.clj:198)
	at compojure.core$routes$fn__103264$f__103265$respond_SINGLEQUOTE___103266.invoke(core.clj:197)
	at compojure.core$wrap_route_matches$fn__103245.invoke(core.clj:152)
	at compojure.core$routes$fn__103264$f__103265.invoke(core.clj:198)
	at compojure.core$routes$fn__103264$f__103265$respond_SINGLEQUOTE___103266.invoke(core.clj:197)
	at compojure.core$wrap_route_matches$fn__103245.invoke(core.clj:152)
	at compojure.core$routes$fn__103264$f__103265.invoke(core.clj:198)
	at compojure.core$routes$fn__103264$f__103265$respond_SINGLEQUOTE___103266.invoke(core.clj:197)
	at compojure.core$wrap_route_matches$fn__103245.invoke(core.clj:152)
	at compojure.core$routes$fn__103264$f__103265.invoke(core.clj:198)
	at compojure.core$routes$fn__103264$f__103265$respond_SINGLEQUOTE___103266.invoke(core.clj:197)
	at metabase.api.util.handlers$_route_map_handler$fn__103327.invoke(handlers.clj:24)
	at metabase.api.open_api.HandlerWithOpenAPISpec.invoke(open_api.clj:42)
	at metabase.api.util.handlers$_route_map_handler$fn__103327.invoke(handlers.clj:23)
	at metabase.api.open_api.HandlerWithOpenAPISpec.invoke(open_api.clj:42)
	at clojure.lang.Var.invoke(Var.java:395)
	at compojure.core$routes$fn__103264$f__103265.invoke(core.clj:198)
	at compojure.core$routes$fn__103264.invoke(core.clj:200)
	at metabase.server.middleware.exceptions$catch_uncaught_exceptions$fn__117359.invoke(exceptions.clj:84)
	at metabase.server.middleware.exceptions$catch_api_exceptions$fn__117356.invoke(exceptions.clj:73)
	at metabase.server.middleware.log$log_api_call$fn__115404$fn__115405$fn__115406.invoke(log.clj:225)
	at metabase.driver.sql_jdbc.execute.diagnostic$do_with_diagnostic_info.invokeStatic(diagnostic.clj:17)
	at metabase.driver.sql_jdbc.execute.diagnostic$do_with_diagnostic_info.invoke(diagnostic.clj:12)
	at metabase.server.middleware.log$log_api_call$fn__115404$fn__115405.invoke(log.clj:216)
	at toucan2.execute$do_with_call_counts.invokeStatic(execute.clj:112)
	at toucan2.execute$do_with_call_counts.invoke(execute.clj:103)
	at metabase.server.middleware.log$log_api_call$fn__115404.invoke(log.clj:215)
	at metabase.server.middleware.browser_cookie$ensure_browser_id_cookie$fn__117587.invoke(browser_cookie.clj:40)
	at metabase.server.middleware.security$add_security_headers$fn__117322.invoke(security.clj:333)
	at metabase.server.middleware.json$wrap_json_body$fn__117014.invoke(json.clj:128)
	at metabase.server.middleware.offset_paging$handle_paging$fn__117447.invoke(offset_paging.clj:48)
	at metabase.server.middleware.json$wrap_streamed_json_response$fn__117002.invoke(json.clj:83)
	at metabase.server.middleware.metadata_provider_cache$wrap_metadata_provider_cache$fn__117597.invoke(metadata_provider_cache.clj:11)
	at ring.middleware.keyword_params$wrap_keyword_params$fn__117862.invoke(keyword_params.clj:55)
	at ring.middleware.params$wrap_params$fn__117881.invoke(params.clj:77)
	at metabase.server.middleware.misc$maybe_set_site_url$fn__114439.invoke(misc.clj:78)
	at metabase.server.middleware.session$reset_session_timeout$fn__117575.invoke(session.clj:299)
	at metabase.server.middleware.session$bind_current_user$fn__117568$fn__117569.invoke(session.clj:274)
	at metabase.request.session$do_with_current_user$fn__87588.invoke(session.clj:61)
	at metabase.request.session$do_with_current_user$do_with_user_local_values__87583.invoke(session.clj:57)
	at metabase.request.session$do_with_current_user.invokeStatic(session.clj:58)
	at metabase.request.session$do_with_current_user.invoke(session.clj:38)
	at metabase.server.middleware.session$bind_current_user$fn__117568.invoke(session.clj:273)
	at metabase.server.middleware.session$wrap_current_user_info$fn__117564.invoke(session.clj:249)
	at metabase.server.middleware.premium_features_cache$wrap_premium_features_cache_check$fn__117425.invoke(premium_features_cache.clj:57)
	at metabase.server.middleware.settings_cache$wrap_settings_cache_check$fn__117392.invoke(settings_cache.clj:50)
	at metabase.analytics.sdk$embedding_mw$embedding_mw_fn__68229.invoke(sdk.clj:69)
	at metabase.server.middleware.session$wrap_session_key$fn__117509.invoke(session.clj:90)
	at metabase.server.middleware.auth$wrap_static_api_key$fn__116958.invoke(auth.clj:17)
	at ring.middleware.cookies$wrap_cookies$fn__117782.invoke(cookies.clj:200)
	at metabase.server.middleware.misc$add_version$fn__114402.invoke(misc.clj:29)
	at metabase.server.middleware.misc$add_content_type$fn__114410.invoke(misc.clj:47)
	at metabase.server.middleware.misc$disable_streaming_buffering$fn__114447.invoke(misc.clj:94)
	at ring.middleware.gzip$wrap_gzip$fn__117824.invoke(gzip.clj:86)
	at metabase.server.middleware.request_id$wrap_request_id$fn__114236.invoke(request_id.clj:10)
	at metabase.server.middleware.misc$bind_request$fn__114450$fn__114451.invoke(misc.clj:106)
	at metabase.request.current$do_with_current_request.invokeStatic(current.clj:19)
	at metabase.request.current$do_with_current_request.invoke(current.clj:15)
	at metabase.server.middleware.misc$bind_request$fn__114450.invoke(misc.clj:105)
	at metabase.server.middleware.ssl$redirect_to_https_middleware$fn__117611.invoke(ssl.clj:52)
	at metabase.server.instance$async_proxy_handler$fn__115108.invoke(instance.clj:82)
	at metabase.server.instance.proxy$org.eclipse.jetty.ee9.servlet.ServletHandler$ff19274a.doHandle(Unknown Source)
	at org.eclipse.jetty.ee9.nested.ScopedHandler.handle(ScopedHandler.java:125)
	at metabase.server.statistics_handler$new_handler$fn__107112.invoke(statistics_handler.clj:105)
	at metabase.server.statistics_handler.proxy$org.eclipse.jetty.ee9.nested.HandlerWrapper$ff19274a.handle(Unknown Source)
	at org.eclipse.jetty.ee9.nested.HandlerWrapper.handle(HandlerWrapper.java:124)
	at org.eclipse.jetty.ee9.nested.ScopedHandler.nextHandle(ScopedHandler.java:195)
	at org.eclipse.jetty.ee9.nested.ContextHandler.doHandle(ContextHandler.java:1041)
	at org.eclipse.jetty.ee9.nested.ScopedHandler.nextScope(ScopedHandler.java:162)
	at org.eclipse.jetty.ee9.servlet.ServletHandler.doScope(ServletHandler.java:483)
	at metabase.server.instance.proxy$org.eclipse.jetty.ee9.servlet.ServletHandler$ff19274a.doScope(Unknown Source)
	at org.eclipse.jetty.ee9.nested.ScopedHandler.nextScope(ScopedHandler.java:160)
	at org.eclipse.jetty.ee9.nested.ContextHandler.doScope(ContextHandler.java:962)
	at org.eclipse.jetty.ee9.nested.ScopedHandler.handle(ScopedHandler.java:123)
	at org.eclipse.jetty.ee9.nested.ContextHandler.handle(ContextHandler.java:1723)
	at org.eclipse.jetty.ee9.nested.HttpChannel$RequestDispatchable.dispatch(HttpChannel.java:1581)
	at org.eclipse.jetty.ee9.nested.HttpChannel.dispatch(HttpChannel.java:733)
	at org.eclipse.jetty.ee9.nested.HttpChannel.handle(HttpChannel.java:520)
	at org.eclipse.jetty.ee9.nested.ContextHandler$CoreContextHandler$CoreToNestedHandler.handle(ContextHandler.java:3048)
	at org.eclipse.jetty.server.handler.ContextHandler.handle(ContextHandler.java:1220)
	at org.eclipse.jetty.server.Server.handle(Server.java:195)
	at org.eclipse.jetty.server.internal.HttpChannelState$HandlerInvoker.run(HttpChannelState.java:680)
	at org.eclipse.jetty.server.internal.HttpConnection.onFillable(HttpConnection.java:411)
	at org.eclipse.jetty.server.internal.HttpConnection$FillableCallback.succeeded(HttpConnection.java:1809)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
	at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:54)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.runTask(AdaptiveExecutionStrategy.java:492)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.epcRunTask(AdaptiveExecutionStrategy.java:428)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.consumeTask(AdaptiveExecutionStrategy.java:401)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.tryProduce(AdaptiveExecutionStrategy.java:255)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.run(AdaptiveExecutionStrategy.java:204)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:312)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:1009)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1239)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1194)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: org.eclipse.jgit.errors.NoRemoteRepositoryException: origin: not found.
	at org.eclipse.jgit.transport.TransportLocal$1.open(TransportLocal.java:100)
	at org.eclipse.jgit.transport.TransportBundleFile$1.open(TransportBundleFile.java:74)
	at org.eclipse.jgit.transport.Transport.open(Transport.java:525)
	at org.eclipse.jgit.transport.Transport.open(Transport.java:277)
	at org.eclipse.jgit.transport.Transport.open(Transport.java:247)
	at org.eclipse.jgit.api.LsRemoteCommand.execute(LsRemoteCommand.java:158)
	at org.eclipse.jgit.api.LsRemoteCommand.call(LsRemoteCommand.java:131)
	at org.eclipse.jgit.api.LsRemoteCommand.call(LsRemoteCommand.java:1)
	at metabase_enterprise.remote_sync.source.git$call_remote_command.invokeStatic(git.clj:56)
	... 136 more
```

### Information about your Metabase installation

```JSON
{
  "browser-info": {
    "language": "en-US",
    "platform": "MacIntel",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:148.0) Gecko/20100101 Firefox/148.0",
    "vendor": ""
  },
  "metabase-info": {
    "valid-thru": "2026-12-31T12:00:00Z",
    "current-user-count": 1,
    "application-database": "postgres",
    "settings": {
      "report-timezone": null
    },
    "application-database-details": {
      "database": {
        "name": "PostgreSQL",
        "version": "17.5 (Debian 17.5-1.pgdg120+1)"
      },
      "jdbc-driver": {
        "name": "PostgreSQL JDBC Driver",
        "version": "42.7.8"
      }
    },
    "databases": [
      "postgres",
      "h2"
    ],
    "max-users": 50,
    "plan-alias": "enterprise-airgap",
    "hosting-env": "unknown",
    "run-mode": "prod",
    "version": {
      "date": "2026-03-02",
      "tag": "v1.59.1",
      "hash": "d609bf8"
    },
    "airgap-token": "enabled"
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "21.0.10+7-LTS",
    "java.vendor": "Eclipse Adoptium",
    "java.vendor.url": "https://adoptium.net/",
    "java.version": "21.0.10",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "21.0.10+7-LTS",
    "os.name": "Linux",
    "os.version": "6.17.8-orbstack-00308-g8f9c941121b1",
    "user.language": "en",
    "user.timezone": "America/Los_Angeles"
  }
}
```

### Severity

P2

### Additional context

_No response_

--- a/enterprise/backend/src/metabase_enterprise/remote_sync/source/git.clj
+++ b/enterprise/backend/src/metabase_enterprise/remote_sync/source/git.clj
@@ -424,7 +424,7 @@
   (FileUtils/deleteDirectory repo-path))
 
 (defn- get-jgit [^File path {:keys [remote-url token] :as args}]
-  (if-let [obj (get @jgit (.getPath path))]
+  (if-let [obj (when (.exists path) (get @jgit (.getPath path)))]
     obj
     (get (swap! jgit assoc (.getPath path) (u/prog1 (open-jgit path {:remote-url remote-url
                                                                      :token      token})

--- a/enterprise/backend/test/metabase_enterprise/remote_sync/source/git_test.clj
+++ b/enterprise/backend/test/metabase_enterprise/remote_sync/source/git_test.clj
@@ -451,6 +451,27 @@
                (map :message (git/log repaired-source)))
             "Should be able to fetch after origin repair")))))
 
+(deftest get-jgit-reclones-after-local-repo-deleted-test
+  (testing "GHY-3815: if the cached local clone dir is deleted out from under us, the next
+            operation re-clones instead of returning a stale cached Git instance (which fails
+            permanently with 'origin: not found' until an instance restart)"
+    (mt/with-temp-dir [remote-dir nil]
+      (let [[source remote] (init-source! "master" remote-dir :branches ["branch-1"])
+            remote-url (:remote-url source)
+            local-path (#'git/repo-path {:remote-url remote-url})
+            ^Git cached-git (:git source)]
+        (is (.exists local-path) "Precondition: local clone dir exists after the initial clone")
+        (is (= ["branch-1" "master"] (source.p/branches source))
+            "Precondition: branches works before the dir is deleted")
+        (FileUtils/deleteDirectory local-path)
+        (is (not (.exists local-path)) "Local clone dir is gone")
+        (let [fresh-source (->source! "master" remote)]
+          (is (.exists local-path) "Local clone dir was re-created (re-cloned)")
+          (is (not (identical? cached-git (:git fresh-source)))
+              "A fresh Git instance is returned, not the stale cached one")
+          (is (= ["branch-1" "master"] (source.p/branches fresh-source))
+              "branches works again after the dir was deleted, without an instance restart"))))))
+
 (deftest ^:parallel credentials-provider-test
   (testing "GitHub URL uses x-access-token"
     (let [provider (git/credentials-provider "https://github.com/org/repo.git" "my-token")]

--- a/enterprise/backend/test/metabase_enterprise/remote_sync/source/git_test.clj
+++ b/enterprise/backend/test/metabase_enterprise/remote_sync/source/git_test.clj
@@ -458,7 +458,7 @@
     (mt/with-temp-dir [remote-dir nil]
       (let [[source remote] (init-source! "master" remote-dir :branches ["branch-1"])
             remote-url (:remote-url source)
-            local-path (#'git/repo-path {:remote-url remote-url})
+            ^File local-path (#'git/repo-path {:remote-url remote-url})
             ^Git cached-git (:git source)]
         (is (.exists local-path) "Precondition: local clone dir exists after the initial clone")
         (is (= ["branch-1" "master"] (source.p/branches source))

